### PR TITLE
Fidelity Enhancement

### LINKED ISSUE #299: Make use of maker fidelity bonds
Our current fidelity bond design has various known security bugs. These needs to be fixed with test coverage to ensure the security of the fidelity protocol. 

This issue will mark completion of all immediate and observed bugs in fidelity bonds. Closure of this issue will also include test coverage for the updated designs.

Major updates needed:

- Include fidelity reedem_script in Proof data. Use it to validate that the time lock is correct as claimed.
- Add fidelity verification at DNS. So it can be used to prevent DOS. 
- Add fidelity verification at Taker offerbook. One more layer of validation in case DNS cheats.

### PR DESCRIPTION
**Pull request by Wukong247**
closes #299

### CODE REVIEW

File: src/market/directory.rs [hunk 1]
```diff
@@ -315,32 +446,98 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), Dir
...
+                .is_ok();
+
+            let current_height = wallet.rpc.get_block_count()?;
+            let lock_time_valid =
+                LockTime::from_height(current_height as u32)? < metadata.proof.bond.lock_time;
+
+            if script_pubkey == redeem_script_pubkey
+                && ecdsa_verified
+                && lock_time_valid
+                && cert_hash == calculated_cert_hash
```
Comment (mojoX911) [inline src/market/directory.rs:501]: This is not a good idea from security standpoint. Ideally each check should be done sequentially, one after other. Starting from the least costly check ending with the most costly check. Merging them together can create potential attack like, someone keeps triggering costly secp verification, by just sending garbage tx.

You wanna put the checks one after another, only moving to the next check if the previous passes. Or else Error quickly.

File: src/taker/offers.rs [hunk 2]
```diff
@@ -32,6 +32,19 @@ pub struct OfferAndAddress {
...
+    Post { metadata: Box<OfferMetadata> },
+    Get { makers: u32 },
+    Dummy { url: String },
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct OfferMetadata {
+    url: String,
+    proof: FidelityProof,
+}
```
Comment (mojoX911) [inline src/taker/offers.rs:46]: These structures are used for all the systems. So lets put them somewhere common.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Concept ACK. I like the structure from the first look. Few cleanups can be done. Also needs some more checks at fidelity validation as mentioned below. 

Also the wallet is DNS is actually not needed. You just need the rpc interface, and that can be created without creating a wallet. That will also remove the redundant wallet codes.

Lets modularize all the resulting logic, mostly the validation part and thenew message data. We should have one common validation routine that is used for both Taker and DNS.

@KnowWhoami also made the test framework/mod.rs more flexible in his last PR so that could be helpful to reduce some more test codes.

File: src/maker/server.rs [hunk 3]
```diff
@@ -181,21 +215,26 @@ fn network_bootstrap(
...
 
-        let request_line = format!("POST {}\n", maker_address);
+        let buffer = serde_cbor::ser::to_vec(&request).unwrap();
+        let length = buffer.len() as u64;
+
+        // Write length and buffer to the stream
         if let Err(e) = stream
-            .write_all(request_line.as_bytes())
+            .write_all(&length.to_be_bytes())
+            .and_then(|_| stream.write_all(&buffer))
```
[the remark this answers is not in the document]
Comment (Wukong247) [reply src/maker/server.rs:224]: We need to improve the read and send routines of ours as its still vanilla stream, but ya makes sense to have them enhanced, and use them here.

File: src/maker/server.rs [hunk 4]
```diff
@@ -50,6 +51,20 @@ pub const MIN_CONTRACT_REACTION_TIME: u16 = 48;
 /// E.g., for 1 billion sats (0.01 BTC), a value of 10_000 would result in a 0.1% fee.
 pub const AMOUNT_RELATIVE_FEE_PPB: Amount = Amount::from_sat(10_000_000);
 
+// Structured requests and responses using serde.
+#[derive(Serialize, Deserialize, Debug)]
+enum Request {
```
[the remark this answers is not in the document]
Comment (Wukong247) [reply src/maker/server.rs:56]: taken

File: src/market/directory.rs [hunk 5]
```diff
@@ -27,12 +37,68 @@ use std::{
...
 pub enum DirectoryServerError {
     IO(std::io::Error),
     Net(NetError),
     MutexPossion,
+    Wallet(WalletError),
+    Cbor(serde_cbor::Error),
+    Rpc(bitcoind::bitcoincore_rpc::Error),
+    OutputsIndexError(OutputsIndexError),
+    Locktime(bitcoin::blockdata::locktime::absolute::ConversionError),
+    Secp(bitcoin::secp256k1::Error),
```
[the remark this answers is not in the document]
Comment (Wukong247) [reply src/market/directory.rs:66]: I need to check the error tree once...

File: src/market/directory.rs [hunk 6]
```diff
@@ -315,32 +424,99 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), Dir
...
+                fidelity_redeemscript(&metadata.proof.bond.lock_time, &metadata.proof.bond.pubkey);
+            let address = Address::p2wsh(
+                fidelity_redeem_script.as_script(),
+                bitcoin::network::Network::Regtest,
+            );
+            let script_pubkey = &address.script_pubkey();
+
+            let txid = metadata.proof.bond.outpoint.txid;
+            let transaction = wallet.rpc.get_raw_transaction(&txid, None)?;
+            let tx_out = transaction.tx_out(0)?;
```
[the remark this answers is not in the document]
Comment (Wukong247) [reply src/market/directory.rs:477]: taken

File: src/market/directory.rs [hunk 7]
```diff
@@ -315,32 +446,98 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), Dir
...
+                let mut addresses = directory
+                    .addresses
+                    .write()
+                    .expect("Failed to acquire write lock");
+                addresses.insert(AddressEntry(
+                    metadata.proof.bond.amount.to_sat(),
+                    metadata.url,
+                ));
+            } else {
+                log::warn!("Potentially suspicious maker detected.");
```
[the remark this answers is not in the document]
Comment (Wukong247) [reply src/market/directory.rs:513]: taken

File: src/market/directory.rs [hunk 8]
```diff
@@ -315,32 +446,98 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), Dir
...
+                .verify_ecdsa(
+                    &cert_message,
+                    &metadata.proof.cert_sig,
+                    &metadata.proof.bond.pubkey.inner,
+                )
+                .is_ok();
+
+            let current_height = wallet.rpc.get_block_count()?;
+            let lock_time_valid =
+                LockTime::from_height(current_height as u32)? < metadata.proof.bond.lock_time;
```
[the remark this answers is not in the document]
Comment (Wukong247) [reply src/market/directory.rs:496]: We are verifying the redeem script in the transaction with the derived redeem script, so if there is any discrepancy, we simply do not allow the entry.

File: src/taker/offers.rs (hunk 2 above)
Comment (Wukong247) [reply src/taker/offers.rs:46]: Yes utills..

File: src/market/directory.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (mojoX911) [reply src/market/directory.rs:496]: The `reedem_script` also should be added inside the `cert_hash` to make it part of the commitment structure. The only source of this reedemscript data is the maker itself, we cannot independently fetch from the blockchain. So better to put it in the commitment string that the maker cannot lie about.

[the remark this answers is not in the document]
Comment (Wukong247) [reply src/market/directory.rs:496]: The redeem script can be constructed using the data provided by the maker in the offer, which includes the public key and locktime. If the script_pub generated from this redeem script (using the public key and locktime) doesn't match with the one we get from tx, we will not add the maker to the directory list.

File: src/market/directory.rs (hunk 1 above)
Comment (mojoX911) [reply src/market/directory.rs:501]: Just a pattern. Not a requirement. That also makes visualising the checks in the code easier.

File: src/market/directory.rs [hunk 9]
```diff
@@ -279,10 +331,10 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), Dir
     while !directory.shutdown.load(Relaxed) {
         match listener.accept() {
             Ok((mut stream, addrs)) => {
-                log::debug!("Incoming connection from : {}", addrs);
-                stream.set_read_timeout(Some(Duration::from_secs(20)))?;
-                stream.set_write_timeout(Some(Duration::from_secs(20)))?;
-                handle_client(&mut stream, &directory.clone())?;
+                log::error!("Incoming connection from : {}", addrs);
```
Comment (mojoX911) [inline src/market/directory.rs:334]: why is this a error?

File: src/market/directory.rs [hunk 10]
```diff
@@ -315,32 +367,63 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), Dir
...
+            ) {
+                Ok(_) => {
+                    log::info!("Maker verified successfully.");
+                    directory.addresses.write()?.insert(AddressEntry(
+                        metadata.proof.bond.amount.to_sat(),
+                        metadata.url.clone(),
+                    ));
+                }
+                Err(_) => {
+                    log::warn!("Potentially suspicious maker detected: {:?}", metadata.url);
```
Comment (mojoX911) [inline src/market/directory.rs:399]: let's also log the error? Allows us to see what type of failure happened.

File: src/market/directory.rs [hunk 11]
```diff
@@ -218,9 +263,16 @@ pub fn write_addresses_to_file(
...
-pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), DirectoryServerError> {
+pub fn start_directory_server(
+    directory: Arc<DirectoryServer>,
+    rpc_config: Option<RPCConfig>,
+) -> Result<(), DirectoryServerError> {
     let mut tor_handle = None;
 
+    let rpc_config = rpc_config.unwrap_or_default();
+
+    let wallet = bitcoincore_rpc::Client::try_from(&rpc_config)?;
```
Comment (mojoX911) [inline src/market/directory.rs:274]: Let call it a rpc_client, instead of wallet.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Ack. Few nits and its good to go.

File: src/market/directory.rs (hunk 9 above)
Comment (Wukong247) [reply src/market/directory.rs:334]: Ah!! revert to debug

File: src/market/directory.rs (hunk 10 above)
Comment (Wukong247) [reply src/market/directory.rs:399]: Cool

File: src/market/directory.rs [hunk 12]
```diff
@@ -330,32 +385,63 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) -> Result<(), Dir
...
+            ) {
+                Ok(_) => {
+                    log::info!("Maker verified successfully.");
+                    directory.addresses.write()?.insert(AddressEntry(
+                        metadata.proof.bond.amount.to_sat(),
+                        metadata.url.clone(),
+                    ));
+                }
+                Err(_) => {
+                    log::error!("Potentially suspicious maker detected: {:?}", metadata.url);
```
Comment (mojoX911) [inline src/market/directory.rs:417]: Let's log the error. Would be important to know in case of any random failure.

**Review by mojoX911 — APPROVED** [review]
Ack. Good to go. Merging this as it's big code and can cause conflicts. 

Just one more log error request that can be packaged in any upcoming PR.

**Pull request merged**